apis/v1alpha3/servicebinding_webhook.go

```diff
@@ -0,0 +1,141 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var servicebindinglog = logf.Log.WithName("servicebinding-resource")
+
+func (r *ServiceBinding) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-servicebinding-io-v1alpha3-servicebinding,mutating=true,failurePolicy=fail,sideEffects=None,groups=servicebinding.io,resources=servicebindings,verbs=create;update,versions=v1alpha3,name=mservicebinding.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Defaulter = &ServiceBinding{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *ServiceBinding) Default() {
+	servicebindinglog.Info("default", "namespace", r.Namespace, "name", r.Name)
+
+	if r.Spec.Name == "" {
+		r.Spec.Name = r.Name
+	}
+}
+
+//+kubebuilder:webhook:path=/validate-servicebinding-io-v1alpha3-servicebinding,mutating=false,failurePolicy=fail,sideEffects=None,groups=servicebinding.io,resources=servicebindings,verbs=create;update,versions=v1alpha3,name=vservicebinding.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Validator = &ServiceBinding{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *ServiceBinding) ValidateCreate() error {
+	servicebindinglog.Info("validate create", "namespace", r.Namespace, "name", r.Name)
+	return r.validate().ToAggregate()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *ServiceBinding) ValidateUpdate(old runtime.Object) error {
+	servicebindinglog.Info("validate update", "namespace", r.Namespace, "name", r.Name)
+	// TODO(user): check for immutable fields, if any
+	return r.validate().ToAggregate()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *ServiceBinding) ValidateDelete() error {
+	servicebindinglog.Info("validate delete", "namespace", r.Namespace, "name", r.Name)
+	return r.validate().ToAggregate()
+}
+
+func (r *ServiceBinding) validate() field.ErrorList {
+	errs := field.ErrorList{}
+
+	errs = append(errs, r.Spec.validate(field.NewPath("spec"))...)
+
+	return errs
+}
+
+func (r *ServiceBindingSpec) validate(fldPath *field.Path) field.ErrorList {
+	errs := field.ErrorList{}
+
+	if r.Name == "" {
```

Comment:
The schema is not and can never be the "authoritative API source of truth". It's a schema. Schema are great up to the point they fall over. The service binding spec is the authoritative source of true for this resource. 

The current schema says these fields are required, but required fields can also be empty. Until your suggestion is upstreamed into the authoritative spec, these checks are appropriate. Even once that happens, I'm not sure these checks are inappropriate.

Reply:
Just saying that these checks can be added to CRD schema so that we do not need to add any additional logic here - k8s API server will perform that check for us. Also,

> The current schema says these fields are required, but required fields can also be empty. Until your suggestion is upstreamed into the authoritative spec, these checks are appropriate.

Should not we first then fix https://github.com/servicebinding/spec#resource-type-schema-1 and then work on the webhook logic?